File: solutions/elixir/018-valid-parentheses.ex
# Problem 18: Valid Parentheses (LeetCode 20)
# Difficulty: Easy
# Language: Elixir
# 
defmodule ValidParentheses do
  def is_valid(s) do
    s
    |> String.graphemes()
    |> Enum.reduce([], fn
      ch, stk when ch in ["(", "{", "["] -> [ch | stk]
      ")", ["(" | rest] -> rest
      "]", ["[" | rest] -> rest
      "}", ["{" | rest] -> rest
      _, stk -> ["!" | stk]  # mismatch
    end)
    |> Enum.empty?()
  end
end

File: solutions/typescript/018-valid-parentheses.ts
/**
 * Problem 18: Valid Parentheses (LeetCode 20)
 * Difficulty: Easy
 * Language: TypeScript
 */
function isValid(s: string): boolean {
  const matching: Record<string, string> = { ')': '(', ']': '[', '}': '{' };
  const stack = [...s].reduce<string[]>((stk, ch) => {
    if ('({['.includes(ch)) return [...stk, ch];
    if (stk.length > 0 && stk[stk.length - 1] === matching[ch]) return stk.slice(0, -1);
    return [...stk, ch]; // mismatch marker
  }, []);
  return stack.length === 0;
}
